### PR TITLE
user_external_accounts: Encrypt account data along with auth data, and handle null values correctly. also update the oob migration runner to use the db handle passed to it

### LINKED ISSUE #20030: User External Account encryption: handle null authdata correctly
Right now we will set authdata to an empty string instead of null if we try to encrypt a null authdata, we should fix this to set it to be actually null, and update the OOB migration to recognise null values when checking progress.

### PR DESCRIPTION
**Pull request by arussellsaw**
closes #20030

When enabling encryption on dogfood i noticed that the oob migration wasn't making any progress, @tsenart and i spent some time debugging & found out a few things:

* we were de-nulling values in auth data
* we weren't encrypting account data, for openid connect (like google) we never store account data, only auth data, so we never encrypted anything for those rows, but we should be as there was sensitive info in the account data field.
* the oob migrator wasn't using a database dependency passed to it

### CODE REVIEW

File: internal/database/oob_migrate.go [hunk 1]
```diff
@@ -215,7 +215,7 @@ func (m *ExternalAccountsMigrator) Progress(ctx context.Context) (float64, error
 				CAST(c1.count AS float) / CAST(c2.count AS float)
 			END
 		FROM
-			(SELECT COUNT(*) AS count FROM user_external_accounts WHERE encryption_key_id != '' OR auth_data = '') c1,
+			(SELECT COUNT(*) AS count FROM user_external_accounts WHERE encryption_key_id != '' AND (account_data IS NOT NULL OR auth_data IS NOT NULL)) c1,
```
Comment (asdine) [inline internal/database/oob_migrate.go:218]: Good catch! Didn't think about this case at all 🤦🏼

**Review by asdine — APPROVED** [review]

File: internal/database/oob_migrate.go (hunk 1 above)
Comment (arussellsaw) [reply internal/database/oob_migrate.go:218]: wait i just realised this logic isn't correct

Comment (arussellsaw) [reply internal/database/oob_migrate.go]: @asdine i think this is correct now

**Comment by ryanslade** [comment]
nit: Your commit title is very long. Can you shorten it and move details into the commit message?

**Comment by unknwon** [comment]
> * we weren't encrypting account data, for openid connect (like google) we never store account data, only auth data, so we never encrypted anything for those rows, but we should be as there was sensitive info in the account data field.

This was left out on purpose (on @asdine's original PR) because I think there is no sensitive info in the account data field. Can you provide an example that it could contain sensitive info?

**Comment by tsenart** [comment]
@unknwon: The information in account_data, while not strictly secret for some providers, looked sensitive, so we just defaulted on encrypting it. Additionally, some providers don't set auth_data, only account_data, which is something we learned while dogfooding this in k8s.sgdev.org

**Comment by unknwon** [comment]
> @unknwon: The information in account_data, while not strictly secret for some providers, looked sensitive, so we just defaulted on encrypting it. 

I don't mind we always be on the safe side (i.e. encrypting more than we need), but hard to agree account_data are sensitive at the moment.

>Additionally, some providers don't set auth_data, only account_data, which is something we learned while dogfooding this in k8s.sgdev.org

If we (can only) rely on impersonate of a sudo access token (or similar), there would be no auth_data for the individual user.

**Comment by tsenart** [comment]
> but hard to agree account_data are sensitive at the moment.

The data we inspected in dogfood was for Google OIDC auth provider. It contained stuff like nonces and hashes inside a JSON blob. Perhaps we didn't need to encrypt it, but we were unsure, so we defaulted to doing it.

> If we (can only) rely on impersonate of a sudo access token (or similar), there would be no auth_data for the individual user.

Not sure how this relates to what I wrote.

**Pull request merged**